components/explorer/src/api/websocket.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2020 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types used in WebSocket communication with the explorer service.
+
+use chrono::{DateTime, Utc};
+use exonum::{
+    blockchain::{Block, Schema, TxLocation},
+    crypto::Hash,
+    merkledb::{access::Access, ListProof},
+    runtime::{ExecutionStatus, InstanceId, MethodId},
+};
+use serde_derive::{Deserialize, Serialize};
+
+use std::fmt;
+
+use super::TransactionHex;
+use crate::median_precommits_time;
+
+/// Messages generated by WebSocket clients.
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+#[serde(tag = "type", content = "payload", rename_all = "snake_case")]
+pub enum IncomingMessage {
+    /// Set subscription for websocket connection.
+    SetSubscriptions(Vec<SubscriptionType>),
+    /// Send transaction to the blockchain.
+    Transaction(TransactionHex),
+}
+
+/// Subscription type (new blocks or committed transactions).
+#[derive(Debug, PartialEq, Eq, Hash, Clone, PartialOrd, Ord)]
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum SubscriptionType {
+    /// Subscription to nothing.
+    None,
+    /// Subscription on new blocks.
+    Blocks,
+    /// Subscription on committed transactions.
+    Transactions {
+        /// Optional filter for subscription.
+        filter: Option<TransactionFilter>,
+    },
+}
+
+/// Filter for transactions by service instance and (optionally) method identifier
+/// within the service.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, PartialOrd, Ord)]
+#[derive(Serialize, Deserialize)]
+pub struct TransactionFilter {
+    /// ID of service.
+    pub instance_id: InstanceId,
+    /// Optional ID of transaction in service (if not set, all transaction of service will be sent).
+    pub method_id: Option<MethodId>,
+}
+
+impl TransactionFilter {
+    /// Create new transaction filter.
+    pub fn new(instance_id: InstanceId, method_id: Option<MethodId>) -> Self {
+        Self {
+            instance_id,
+            method_id,
+        }
+    }
+}
+
+/// Response to a WebSocket client.
+#[serde(tag = "result", rename_all = "snake_case")]
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub enum Response<T> {
+    /// Successful response.
+    Success {
+        /// Payload attached to the response.
+        response: T,
+    },
+    /// Erroneous response.
```

Comment:
Nit: I'm not sure that this response can be called "erroneous", the response *itself* is successful (if it was sent, received and parsed), but it rather contains a report about an occurred error.